src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+        // 'true' if the matched type catches all of them, 'false' if it catches none of them, and
+        // 'null' if it might catch some of them. For this test we assume the expression's value
+        // isn't null.
+        public static bool? ExpressionOfTypeMatchesPatternType(this Conversions conversions, TypeSymbol expressionType, TypeSymbol patternType)
+        {
+            if (expressionType == patternType) return true;
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            var conversion = conversions.ClassifyConversion(expressionType, patternType, ref useSiteDiagnostics, builtinOnly: true);
+            // This is for classification purposes only; we discard use-site diagnostics. Use-site diagnostics will
+            // be given if a conversion is actually used.
+            switch (conversion.Kind)
+            {
+                case ConversionKind.Boxing:             // a value of type int matches a pattern of type object
+                case ConversionKind.Identity:           // a value of a given type matches a pattern of that type
+                case ConversionKind.ImplicitReference:  // a value of type string matches a pattern of type object
+
+                case ConversionKind.ImplicitNullable:   // a value of type int matches a pattern of type int?
+                case ConversionKind.ExplicitNullable:   // a non-null value of type "int?" matches a pattern of type int
+                    // but if the types differ (e.g. one of them is type byte and the other is type int?).. no match
+                    return expressionType.StrippedType().TupleUnderlyingType == patternType.StrippedType().TupleUnderlyingType;
+
+                case ConversionKind.ExplicitEnumeration:// a value of enum type does not match a pattern of integral type
+                case ConversionKind.ExplicitNumeric:    // a value of type long does not match a pattern of type int
+                case ConversionKind.ImplicitNumeric:    // a value of type short does not match a pattern of type int
+                case ConversionKind.ImplicitTuple:      // distinct tuple types don't match
+                case ConversionKind.NoConversion:
+                    return false;
+
+                case ConversionKind.ExplicitDynamic:    // a value of type dynamic might not match a pattern of type other than object
+                case ConversionKind.ExplicitReference:  // a narrowing reference conversion might or might not succeed
+                case ConversionKind.Unboxing:           // a value of type object might match a pattern of type int
+                    return null;
+
+                default:
+                    // other conversions don't apply (e.g. conversions from expression, user-defined) and should not arise
+                    throw ExceptionUtilities.UnexpectedValue(conversion.Kind);
+            }
+        }
+    }
+
+    internal abstract class DecisionTree
```

Comment:
It feels like the code in this file is very complicated. At the same time I don't have any alternative to offer at the moment. At the very least, I think all members of DecisionTree classes and the types themselves should have good detailed comments. What do they represent, when and maybe why that information is important. Also, consider describing what problem we are trying to solve and the principles of building the decision tree. For example, what do we start with, how each particular decision handles addition, why things get added to Default before adding to decisions, etc. If there are some invariants that should be maintained in the tree, consider listing those as well. All this information will be very helpful to the next developer that will be tasked with fixing a bug in this component. After reading all this code, I do not pretend to understand what is going on here.